File: common/src/main/java/com/vladte/devhack/common/service/domain/InterviewStageCategoryService.java
package com.vladte.devhack.common.service.domain;

import com.vladte.devhack.entities.InterviewStageCategory;

import java.util.Optional;
import java.util.UUID;

/**
 * Service interface for InterviewStageCategory entity operations.
 */
public interface InterviewStageCategoryService extends BaseService<InterviewStageCategory, UUID> {

    /**
     * Find interview stage category by code.
     *
     * @param code the category code
     * @return optional interview stage category
     */
    Optional<InterviewStageCategory> findByCode(String code);
}

File: angular-ui/src/app/services/global/interview-stage-category.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {InterviewStageCategoryDTO} from '../../models/global/interview-stage-category.model';
import {BaseService} from '../base.service';

/**
 * Service for managing interview stage categories.
 * Extends BaseService to inherit common CRUD operations and follows OOP principles.
 */
@Injectable({
    providedIn: 'root'
})
export class InterviewStageCategoryService extends BaseService<InterviewStageCategoryDTO> {
    protected baseUrl = '/api/interview-stage-categories';

    constructor(http: HttpClient) {
        super(http);
    }

    /**
     * Get interview stage category by code
     * @param code - Category code
     * @returns Observable interview stage category
     */
    getByCode(code: string): Observable<InterviewStageCategoryDTO> {
        return this.http.get<InterviewStageCategoryDTO>(`${this.baseUrl}/by-code/${code}`);
    }
}
